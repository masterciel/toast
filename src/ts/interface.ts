import _Vue from "vue";
import ToastContainer from "../components/VtToastContainer.vue";
import events from "./events";
import { TYPE, EVENTS } from "./constants";
import { getId } from "./utils";
import { ToastContent, ToastOptions, ToastID, PluginOptions } from "../types";

const ToastInterface = (Vue: typeof _Vue, globalOptions?: PluginOptions) => {
  new (Vue.extend(ToastContainer))({
    el: document.createElement("div"),
    propsData: globalOptions
  });
  /**
   * Display a toast
   */
  const toast = (content: ToastContent, options?: ToastOptions): ToastID => {
    const props = Object.assign(
      {},
      { id: getId(), type: TYPE.DEFAULT },
      options,
      { content }
    );
    events.$emit(EVENTS.ADD, props);
    return props.id;
  };
  /**
   * Clear all toasts
   */
  toast.clear = () => events.$emit(EVENTS.CLEAR);
  /**
   * Update Plugin Defaults
   */
  toast.updateDefaults = (update: PluginOptions) => {
    events.$emit(EVENTS.UPDATE_DEFAULTS, update);
  };
  /**
   * Dismiss toast specified by an id
   */
  toast.dismiss = (id: ToastID) => {
    events.$emit(EVENTS.DISMISS, id);
  };
  /**
   * Update Toast
   */
  toast.update = (
    id: ToastID,
    { content, options }: { content?: ToastContent; options?: ToastOptions },
    create = false
  ) => {
    events.$emit(EVENTS.UPDATE, {
      id,
      options: Object.assign({}, options, { content }),
      create
    });
  };
  /**
   * Display a success toast
   */
  toast.success = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.SUCCESS }
  ) => toast(content, Object.assign({}, options, { type: TYPE.SUCCESS }));

  /**
   * Display an info toast
   */
  toast.info = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.INFO }
  ) => toast(content, Object.assign({}, options, { type: TYPE.INFO }));

  /**
   * Display an error toast
   */
  toast.error = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.ERROR }
  ) => toast(content, Object.assign({}, options, { type: TYPE.ERROR }));

  /**
   * Display a warning toast
   */
  toast.warning = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.WARNING }
  ) => toast(content, Object.assign({}, options, { type: TYPE.WARNING }));

  return toast;
};

export default ToastInterface;
