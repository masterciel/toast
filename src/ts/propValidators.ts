import { PropType } from "vue";
import {
  ToastContent,
  CommonOptions,
  PluginOptions,
  ToastID,
  ToastOptions
} from "../types";
import { TYPE, POSITION, VT_NAMESPACE } from "./constants";

const COMMON = {
  type: {
    type: String as PropType<TYPE>,
    default: TYPE.DEFAULT
  },
  classNames: {
    type: [String, Array] as PropType<string | string[]>,
    default: () => []
  },
  trueBoolean: {
    type: Boolean,
    default: true
  }
};

const ICON = {
  type: COMMON.type,
  customIcon: {
    type: [String, Boolean, Object, Function] as PropType<
      NonNullable<CommonOptions["icon"]>
    >,
    default: true
  }
};

const CLOSE_BUTTON = {
  component: {
    type: [String, Object, Function, Boolean] as PropType<
      NonNullable<CommonOptions["closeButton"]>
    >,
    default: "button"
  },
  classNames: COMMON.classNames,
  showOnHover: Boolean,
  ariaLabel: {
    type: String,
    default: "close"
  }
};

const PROGRESS_BAR = {
  timeout: {
    type: [Number, Boolean] as PropType<number | false>,
    default: 5000
  },
  hideProgressBar: Boolean,
  isRunning: Boolean
};

const TRANSITION = {
  transition: {
    type: [Object, String] as PropType<
      NonNullable<PluginOptions["transition"]>
    >,
    default: `${VT_NAMESPACE}__bounce`
  },
  transitionDuration: {
    type: [Number, Object] as PropType<
      NonNullable<PluginOptions["transitionDuration"]>
    >,
    default: 750
  }
};

const CORE_TOAST = {
  position: {
    type: String as PropType<POSITION>,
    default: POSITION.TOP_RIGHT
  },
  draggable: COMMON.trueBoolean,
  draggablePercent: {
    type: Number,
    default: 0.6
  },
  pauseOnFocusLoss: COMMON.trueBoolean,
  pauseOnHover: COMMON.trueBoolean,
  closeOnClick: COMMON.trueBoolean,
  timeout: PROGRESS_BAR.timeout,
  hideProgressBar: PROGRESS_BAR.hideProgressBar,
  toastClassName: COMMON.classNames,
  bodyClassName: COMMON.classNames,
  icon: ICON.customIcon,
  closeButton: CLOSE_BUTTON.component,
  closeButtonClassName: CLOSE_BUTTON.classNames,
  showCloseButtonOnHover: CLOSE_BUTTON.showOnHover,
  accessibility: {
    type: Object as PropType<PluginOptions["accessibility"]>,
    default: () => ({
      toastRole: "alert",
      closeButtonLabel: "close"
    })
  },
  rtl: Boolean
};

const TOAST = {
  id: {
    type: [String, Number] as PropType<ToastID>,
    required: true
  },
  type: COMMON.type,
  content: {
    type: [String, Object, Function] as PropType<ToastContent>,
    required: true
  },
  onClick: Function as PropType<NonNullable<ToastOptions["onClick"]>>,
  onClose: Function as PropType<NonNullable<ToastOptions["onClose"]>>
};

const CONTAINER = {
  newestOnTop: COMMON.trueBoolean,
  maxToasts: {
    type: Number,
    default: 20
  },
  container: {
    type: (Element as unknown) as PropType<
      NonNullable<PluginOptions["container"]>
    >,
    default: () => document.body
  },
  filterBeforeCreate: Function as PropType<
    NonNullable<PluginOptions["filterBeforeCreate"]>
  >,
  filterToasts: Function as PropType<
    NonNullable<PluginOptions["filterToasts"]>
  >,
  containerClassName: COMMON.classNames
};

export default {
  CORE_TOAST,
  TOAST,
  CONTAINER,
  PROGRESS_BAR,
  ICON,
  TRANSITION,
  CLOSE_BUTTON
};
