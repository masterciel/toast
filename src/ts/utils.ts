import { Component, defineComponent } from "vue"
import type {
  ToastComponent,
  ToastContent,
  RenderableToastContent,
} from "../types"

interface DictionaryLike {
  [index: string]: unknown
}

// eslint-disable-next-line @typescript-eslint/ban-types
const isFunction = (value: unknown): value is Function =>
  typeof value === "function"

const isString = (value: unknown): value is string => typeof value === "string"

const isNonEmptyString = (value: unknown): value is string =>
  isString(value) && value.trim().length > 0

const isNumber = (value: unknown): value is number => typeof value === "number"

const isUndefined = (value: unknown): value is undefined =>
  typeof value === "undefined"

const isObject = (value: unknown): value is DictionaryLike =>
  typeof value === "object" && value !== null

const isJSX = (obj: unknown): obj is JSX.Element =>
  hasProp(obj, "tag") && isNonEmptyString(obj.tag)

const isTouchEvent = (event: Event): event is TouchEvent =>
  window.TouchEvent && event instanceof TouchEvent

const isToastComponent = (obj: unknown): obj is ToastComponent =>
  hasProp(obj, "component") && isToastContent(obj.component)

const isVueComponent = (c: unknown): c is Component => {
  if (isFunction(c)) {
    return true
  }
  return isObject(c)
}

const isToastContent = (obj: unknown): obj is ToastContent =>
  // Ignore undefined
  !isUndefined(obj) &&
  // Is a string
  (isString(obj) ||
    // Regular Vue component
    isVueComponent(obj) ||
    // Nested object
    isToastComponent(obj))

const isDOMRect = (obj: unknown): obj is DOMRect =>
  isObject(obj) &&
  ["height", "width", "right", "left", "top", "bottom"].every(p =>
    isNumber(obj[p])
  )

const hasProp = <O extends unknown, K extends PropertyKey>(
  obj: O,
  propKey: K
): obj is O & { [key in K]: unknown } =>
  Object.prototype.hasOwnProperty.call(obj, propKey)

/**
 * ID generator
 */
const getId = (i => () => i++)(0)

function getX(event: MouseEvent | TouchEvent) {
  return isTouchEvent(event) ? event.targetTouches[0].clientX : event.clientX
}

function getY(event: MouseEvent | TouchEvent) {
  return isTouchEvent(event) ? event.targetTouches[0].clientY : event.clientY
}

const removeElement = (el: Element) => {
  if (!isUndefined(el.remove)) {
    el.remove()
  } else if (el.parentNode) {
    el.parentNode.removeChild(el)
  }
}

const getVueComponentFromObj = (obj: ToastContent): RenderableToastContent => {
  if (isToastComponent(obj)) {
    // Recurse if component prop
    return getVueComponentFromObj(obj.component)
  }
  if (isJSX(obj)) {
    // Create render function for JSX
    return defineComponent({
      render() {
        return obj
      },
    })
  }
  // Return the actual object if regular vue component
  return obj
}

export {
  getId,
  getX,
  getY,
  removeElement,
  isString,
  isNonEmptyString,
  isToastContent,
  getVueComponentFromObj,
  hasProp,
  isUndefined,
  isDOMRect,
  isFunction,
}
