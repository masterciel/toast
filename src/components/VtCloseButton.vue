<template>
  <component
    :is="buttonComponent"
    :aria-label="ariaLabel"
    :class="classes"
    v-bind="$attrs"
  >
    &times;
  </component>
</template>
<script lang="ts">
import { computed, defineComponent } from "vue"

import { VT_NAMESPACE } from "../ts/constants"
import PROPS from "../ts/propValidators"
import { getVueComponentFromObj } from "../ts/utils"

export default defineComponent({
  name: "VtCloseButton",

  props: PROPS.CLOSE_BUTTON,

  setup(props) {
    const buttonComponent = computed(() => {
      if (props.component !== false) {
        return getVueComponentFromObj(props.component)
      }
      return "button"
    })
    const classes = computed(() => {
      const classes = [`${VT_NAMESPACE}__close-button`]
      if (props.showOnHover) {
        classes.push("show-on-hover")
      }
      return classes.concat(props.classNames)
    })
    return {
      buttonComponent,
      classes,
    }
  },
})
</script>
