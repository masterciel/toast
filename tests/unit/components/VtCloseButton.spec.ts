import { mount } from "@vue/test-utils"
import VtCloseButton from "../../../src/components/VtCloseButton.vue"
import { VT_NAMESPACE } from "../../../src/ts/constants"
import Simple from "../../utils/components/Simple.vue"

describe("VtCloseButton", () => {
  it("matches default snapshot", () => {
    const wrapper = mount(VtCloseButton, {
      props: {
        component: false,
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })
  it("has default class", () => {
    const wrapper = mount(VtCloseButton, {
      props: {
        component: false,
      },
    })
    expect(wrapper.classes()).toContain(`${VT_NAMESPACE}__close-button`)
  })
  it("is a button by default", () => {
    const wrapper = mount(VtCloseButton, {
      props: {
        component: false,
      },
    })
    expect(wrapper.element.tagName).toEqual("BUTTON")
  })
  it("string custom component", () => {
    const wrapper = mount(VtCloseButton, {
      props: {
        component: "div",
      },
    })
    expect(wrapper.element.tagName).toEqual("DIV")
    expect(wrapper.element).toMatchSnapshot()
  })
  it("vue custom component", () => {
    const wrapper = mount(VtCloseButton, {
      props: {
        component: Simple,
      },
    })
    expect(wrapper.findComponent(Simple).element).toBeTruthy()
    expect(wrapper.element).toMatchSnapshot()
  })
  it("adds 'show-on-hover' class", () => {
    const wrapper = mount(VtCloseButton, {
      props: {
        component: false,
        showOnHover: false,
      },
    })
    const wrapper2 = mount(VtCloseButton, {
      props: {
        component: false,
        showOnHover: true,
      },
    })
    expect(wrapper.classes()).not.toContain("show-on-hover")
    expect(wrapper2.classes()).toContain("show-on-hover")
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper2.element).toMatchSnapshot()
  })
  it("adds custom class string", () => {
    const wrapper = mount(VtCloseButton, {
      props: {
        component: false,
        classNames: "my-class",
      },
    })
    expect(wrapper.classes()).toContain("my-class")
    expect(wrapper.element).toMatchSnapshot()
  })
  it("adds custom class array", () => {
    const wrapper = mount(VtCloseButton, {
      props: {
        component: false,
        classNames: ["my-class", "my-class2"],
      },
    })
    expect(wrapper.classes()).toContain("my-class")
    expect(wrapper.classes()).toContain("my-class2")
    expect(wrapper.element).toMatchSnapshot()
  })
  it("attaches onClick listener", () => {
    const onClick = jest.fn()
    const wrapper = mount(VtCloseButton, {
      props: {
        component: false,
      },
      attrs: { onClick },
    })
    expect(onClick).not.toHaveBeenCalled()
    wrapper.trigger("click")
    expect(onClick).toHaveBeenCalled()
  })
  it("renders default aria label", () => {
    const wrapper = mount(VtCloseButton)
    expect(wrapper.find("button[aria-label='close']").exists()).toBe(true)
    expect(wrapper.element).toMatchSnapshot()
  })
  it("renders custom aria label", () => {
    const wrapper = mount(VtCloseButton, {
      props: {
        ariaLabel: "text",
      },
    })
    expect(wrapper.find("button[aria-label='text']").exists()).toBe(true)
    expect(wrapper.element).toMatchSnapshot()
  })
})
