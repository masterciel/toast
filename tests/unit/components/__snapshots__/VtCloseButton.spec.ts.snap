// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VtCloseButton adds 'show-on-hover' class 1`] = `
<button
  aria-label="close"
  class="Vue-Toastification__close-button"
>
  
  ×

</button>
`;

exports[`VtCloseButton adds 'show-on-hover' class 2`] = `
<button
  aria-label="close"
  class="Vue-Toastification__close-button show-on-hover"
>
  
  ×

</button>
`;

exports[`VtCloseButton adds custom class array 1`] = `
<button
  aria-label="close"
  class="Vue-Toastification__close-button my-class my-class2"
>
  
  ×

</button>
`;

exports[`VtCloseButton adds custom class string 1`] = `
<button
  aria-label="close"
  class="Vue-Toastification__close-button my-class"
>
  
  ×

</button>
`;

exports[`VtCloseButton matches default snapshot 1`] = `
<button
  aria-label="close"
  class="Vue-Toastification__close-button"
>
  
  ×

</button>
`;

exports[`VtCloseButton renders custom aria label 1`] = `
<button
  aria-label="text"
  class="Vue-Toastification__close-button"
>
  
  ×

</button>
`;

exports[`VtCloseButton renders default aria label 1`] = `
<button
  aria-label="close"
  class="Vue-Toastification__close-button"
>
  
  ×

</button>
`;

exports[`VtCloseButton string custom component 1`] = `
<div
  aria-label="close"
  class="Vue-Toastification__close-button"
>
  
  ×

</div>
`;

exports[`VtCloseButton vue custom component 1`] = `
<div
  aria-label="close"
  class="Vue-Toastification__close-button"
>
  Example component
</div>
`;
